### DISCUSSION TITLE
How to properly types thing when using immer middleware?

### DISCUSSION THREAD

**Discussion by rzec-allma** [pmndrs/zustand#2012] [Q&A] [answered]
When using the immer middleware, typing it in typescript seem to be an issue in some cases. For example, I have this code:

```ts
const createReactHelpers = <TCreateOptions, TStore, TState>(createStore: () => StoreApi<TStore>) => {
  // ...
}
```

While this sometimes works when passing a `createStore()` that uses immer, something I this this error:
![Screenshot 2023-08-26 at 8 45 38 AM](https://github.com/pmndrs/zustand/assets/93214391/0f7710be-a0ec-49ef-a0d9-f2c8541d9bd7)

While I would think wrapping the type with `WithImmer` would fix it, the issue is that is seems like none of the immer based type are exported. 

Why are they not exported?
How can I type something like this?

**Comment by dai-shi** [comment]
Can you please create a reproduction with https://tsplay.dev/Wkllpw ?

**Comment by rzec-allma** [comment]
@dai-shi Was not sure how to add imports to that playground so here is a codesandbox with the issue.

I narrowed it down to it seeming to be dependent on the data that is being stored in the store. In this example, the issue is the fact I am storing an `Auth0Client` in the store. If you change that type to `string` or something, the error goes away. Keep in mind, all this code works just fine.

I am working around it by creating my own file that copies the required types for `WithImmer` like this:
```ts
import { Draft } from 'immer';

type Write<T, U> = Omit<T, keyof U> & U;
type SkipTwo<T> = T extends {
	length: 0;
}
	? []
	: T extends {
			length: 1;
	  }
	? []
	: T extends {
			length: 0 | 1;
	  }
	? []
	: T extends [unknown, unknown, ...infer A]
	? A
	: T extends [unknown, unknown?, ...infer A]
	? A
	: T extends [unknown?, unknown?, ...infer A]
	? A
	: never;
type StoreImmer<S> = S extends {
	getState: () => infer T;
	setState: infer SetState;
}
	? SetState extends (...a: infer A) => infer Sr
		? {
				setState(
					nextStateOrUpdater: T | Partial<T> | ((state: Draft<T>) => void),
					shouldReplace?: boolean | undefined,
					...a: SkipTwo<A>
				): Sr;
		  }
		: never
	: never;

export type WithImmer<S> = Write<S, StoreImmer<S>>;
```

And then doing:
```ts
const createReactHelpers = <TCreateOptions, TStore, TState>(createStore: () => WithImmer<StoreApi<TStore>>) => {
```

 instead of:
```ts
const createReactHelpers = <TCreateOptions, TStore, TState>(createStore: () => StoreApi<TStore>) => {
```

If the `WithImmer` type was exported by the package itself, it would avoid me having to do that which would break if those types changed (I would think it is not likely but still...).

**Reply by dai-shi** [reply]
If `WithImmer` type changes, your code will break anyway, but I understand that we don't want to "copy" the type.

It has discussed before, and we consider `WithFoo` is an internal type only to use for overloads. Basically, the solution is to use `typeof` a concrete value, or type things manually for the usage. (I know it's annoying, but it should be the right solution.)

**Reply by rzec-allma** [reply]
Well I am not going to create an instance of every store jsut so typescript does not have a hissy fit and right now `createReactHelpers()` is about 40 lines of code that would have to be duplicated for every single store to do manually (and god forbid I decide to add something in that part of the code, that is why it is abstracted).

If you don't think that type should be exported, that is your call, that just leaves me with have to manually create the type and update it when needed.

**Reply by dai-shi** [reply, marked as answer]
The helper function for abstraction is good. You should manually type it, but I don't mean to copy `WithImmer`. I would just type what I used in the function. So, it depends how 40 lines look like. See: https://github.com/pmndrs/zustand/discussions/2012#discussioncomment-6847141

**Reply by rzec-allma** [reply]
Sorry, did not notice the update playground and did not think about typing that with an extends which does seem to work so thanks.

**Comment by rzec-allma** [comment]
@dai-shi Forgot the link: https://codesandbox.io/s/typescript-playground-export-forked-vgcfvh?file=/index.ts

**Reply by dai-shi** [reply]
You can do imports in the playground: https://tsplay.dev/Wyz2xw

It depends on what you would like in `createReactHelpers`, but here's one example: https://tsplay.dev/W4VJ7N
